plugins/yjs/fps_yjs/routes.py

```
@@ -405,25 +412,34 @@ async def maybe_clean_room(self, room, ws_path: str) -> None:
         file_id = ws_path.split(":", 2)[2]
         # keep the document for a while in case someone reconnects
         await sleep(self.config.document_cleanup_delay)
-        document = self.documents[ws_path]
-        document.unobserve()
-        del self.documents[ws_path]
-        documents = [v for k, v in self.documents.items() if k.split(":", 2)[2] == file_id]
-        if not documents:
-            self.watchers[file_id].cancel(raise_exception=False)
-            await self.watchers[file_id].wait()
-            if file_id in self.watchers:
-                del self.watchers[file_id]
-        room_name = self.websocket_server.get_room_name(room)
-        self.websocket_server.delete_room(room=room)
-
-        if ws_path in self.room_write_permissions:
-            del self.room_write_permissions[ws_path]
 
-        file_path = await self.get_file_path(file_id, document)
-        logger.info("Closing collaboration room", room_id=room_name, file_path=file_path)
-        if room in self.cleaners:
-            del self.cleaners[room]
+        async with self.room_lock(ws_path):
+            try:
+                # If for some reason this room has already been cleaned, do nothing
```

What could be a reason for the room to already be cleaned?

---

Just in case there are other concurrency issues that aren't accounted for. Although maybe it would be better to fail loudly instead?

---

This PR should ensure that only one leaving client can clean the room, so it should never happen.